### ISSUE TITLE
Treat accesses to volatile values as relaxed atomics

### ISSUE THREAD

**Issue by OlivierNicole** [ocaml-multicore/ocaml-tsan#25]
An interesting discussion started at https://github.com/ocaml/ocaml/pull/12030 about avoiding false positives in the OCaml runtime by treating `volatile` values as more or less relaxed atomics. @gasche told us about the existence of the `tsan-distinguish-volatile` flag which might allow just that.

Putting aside portability difficulties between gcc and clang, there may be a possibility to make it work by implementing the instrumentation primitive `__tsan_volatile_read8` as such:
```c
void __tsan_volatile_read8(void *ptr)
{
  __tsan_atomic64_load(ptr, memory_order_relaxed);
}
```
essentially ignoring the result of the atomic read. (For reads other than 8 bytes, it would be the same.)

And I think we can do a similar thing to treat writes to `volatile`s as plain stores.

Edit: added the `memory_order_relaxed` argument, which was missing.

**Comment by gasche** [comment]
I think that the ideal scenario would be for TSan runtimes to expose to external users their `MemoryAccess` function. In fact I don't understand how the `volatile` callbacks are supposed to be implemented without access to this function. It would indeed be simpler if those volatile callbacks used the same type as atomic operation, that is if they performed the access instead of being a pre-access documentation.

From a distance it looks easier to convince TSan upstreams to expose a `memory_access` function than to change the behavior of the volatile support knob, but I am not sure. Maybe they are worried about exposing their internals and they don't care so much about this specific feature.

I think that it would be make sense to ask Marco Elver, the person who implemented volatile support in llvm and clang TSans and also apparently in kcsan. Would you write him an email? (Would you prefer if I did?)

In the meantime, duplicating the operation (once as relaxed then as atomic) should be okay. I think that for stores it could hurt performance a bit, but it's worth trying -- you could compare against the version whose volatile callback does nothing at all.

**Comment by OlivierNicole** [comment]
> From a distance it looks easier to convince TSan upstreams to expose a `memory_access` function than to change the behavior of the volatile support knob, but I am not sure. Maybe they are worried about exposing their internals and they don't care so much about this specific feature.
> 
> I think that it would be make sense to ask Marco Elver, the person who implemented volatile support in llvm and clang TSans and also apparently in kcsan. Would you write him an email? (Would you prefer if I did?)

Yes, it makes sense to expose our need for such functions. I will contact him.

**Comment by OlivierNicole** [comment]
I finally found the time to implement the approach described in my initial message in this issue. After re-reading https://github.com/ocaml/ocaml/issues/10992 once again, I realized that it should in fact suffice to remove our false positives related to `volatile` in TSan, as `volatile` is used to avoid races when using macros for *reading*.

From what I can see so far looking at a few compiler tests, doing this avoids the false positives on `volatile` reads. I still need to confirm this on all tests and if we keep this approach, update https://github.com/ocaml/ocaml/issues/11040 accordingly.
